test/integration/service/service_test.go

```diff
@@ -551,6 +551,209 @@ func Test_TransitionsForTrafficDistribution(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Failed to update status for test pod to Ready: %v", err)
+	}
+	_, err = client.CoreV1().Services(ns.Name).Create(ctx, svc, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to create test service: %v", err)
+	}
+
+	////////////////////////////////////////////////////////////////////////////
+	// Assert that without the presence of `trafficDistribution` field there are
+	// no zone hints in EndpointSlice.
+	////////////////////////////////////////////////////////////////////////////
+
+	assertEndpointSliceHints(t, ctx, client, ns.GetName(), svc.GetName(), false, false)
+
+	////////////////////////////////////////////////////////////////////////////
+	// Update the service by setting `trafficDistribution: PreferSameZone`
+	//
+	// Assert that the respective EndpointSlices get the same-zone hints.
+	////////////////////////////////////////////////////////////////////////////
+
+	trafficDist := corev1.ServiceTrafficDistributionPreferSameZone
+	svc.Spec.TrafficDistribution = &trafficDist
+	_, err = client.CoreV1().Services(ns.Name).Update(ctx, svc, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to update test service with 'trafficDistribution: PreferSameZone': %v", err)
+	}
+
+	assertEndpointSliceHints(t, ctx, client, ns.GetName(), svc.GetName(), true, false)
+
+	////////////////////////////////////////////////////////////////////////////
+	// Change `trafficDistribution` to `PreferSameNode`.
+	//
+	// Assert that the respective EndpointSlices have both same-zone and
+	// same-node hints.
+	////////////////////////////////////////////////////////////////////////////
+
+	trafficDist = corev1.ServiceTrafficDistributionPreferSameNode
+	svc.Spec.TrafficDistribution = &trafficDist
+	_, err = client.CoreV1().Services(ns.Name).Update(ctx, svc, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to update test service with 'trafficDistribution: PreferSameNode': %v", err)
+	}
+
+	assertEndpointSliceHints(t, ctx, client, ns.GetName(), svc.GetName(), true, true)
+
+	////////////////////////////////////////////////////////////////////////////
+	// Remove the field `trafficDistribution` from the service.
+	//
+	// Assert that EndpointSlice for service again has no zone hints.
+	////////////////////////////////////////////////////////////////////////////
+
+	svc.Spec.TrafficDistribution = nil
+	_, err = client.CoreV1().Services(ns.Name).Update(ctx, svc, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to remove annotation 'service.kubernetes.io/topology-mode=Auto' from service: %v", err)
+	}
+
+	assertEndpointSliceHints(t, ctx, client, ns.GetName(), svc.GetName(), false, false)
+
+	////////////////////////////////////////////////////////////////////////////
+	// Update the Node to no longer have a zone label, and re-enable
+	// `PreferSameNode`.
+	//
+	// Assert that the respective EndpointSlices have same-node hints but not
+	// same-zone.
```

Comment:
this was mean :)

Reply:
hm?

It's a valid scenario; we want to make sure that PreferSameNode still works in kind/metal/etc clusters that don't have zones.

Reply:
yeah, I was trying to say that is a really good edge case